Frameworks/Foundation/NSKVOSupport.mm

```diff
@@ -516,169 +515,292 @@ - (void)removeObserver:(id)observer forKeyPath:(NSString*)keyPath {
-            continue;
-        } else {
-            NSKeyValueObservingOptions options = keypathObserver.options;
-            id rootObject = keypathObserver.object;
+        // Skip any keypaths that are in the process of changing.
+        if ([keypathObserver pushWillChange]) {
+            // Call into the lambda function, which will do the actual set-up and notifying if necessary
+            func(keyObserver);
+
+            // This must happen regardless of whether we are currently notifying.
+            _removeNestedObserversAndOptionallyDependents(keyObserver, false);
+        }
+    }
+}
+
+// void TFunc(_NSKVOKeyObserver* keyObserver, NSMutableDictionary* keypathValueCache);
+template <typename TFunc>
+static void _dispatchDidChange(id notifyingObject, NSString* key, TFunc&& func) {
+    _NSKVOObservationInfo* observationInfo = (__bridge _NSKVOObservationInfo*)[notifyingObject observationInfo];
+    NSArray<_NSKVOKeyObserver*>* observers = [observationInfo observersForKey:key];
+    NSMutableDictionary* keypathValueCache = [NSMutableDictionary dictionaryWithCapacity:[observers count]];
+    for (_NSKVOKeyObserver* keyObserver in [observers reverseObjectEnumerator]) {
+        // This must happen regardless of whether we are currently notifying.
+        _addNestedObserversAndOptionallyDependents(keyObserver, false);
+
+        // Skip any keypaths that are in the process of changing.
+        _NSKVOKeypathObserver* keypathObserver = keyObserver.keypathObserver;
+        if ([keypathObserver popDidChange]) {
+            func(keyObserver, keypathValueCache);
+
             id observer = keypathObserver.observer;
             NSString* keypath = keypathObserver.keypath;
+            id rootObject = keypathObserver.object;
+            NSKeyValueObservingOptions options = keypathObserver.options;
+            NSMutableDictionary* change = keypathObserver.pendingChange;
             void* context = keypathObserver.context;
+            [observer observeValueForKeyPath:keypath ofObject:rootObject change:change context:context];
+            keypathObserver.pendingChange = nil;
+        }
+    }
+}
 
-            // The reference platform does not support to-many mutations on nested keypaths.
-            // We have to treat them as to-one mutations to support aggregate functions.
-            if (changeKind != NSKeyValueChangeSetting && keyObserver.restOfKeypathObserver) {
-                // This only needs to be done in willChange because didChange derives from the existing changeset.
-                change[NSKeyValueChangeKindKey] = @(changeKind = NSKeyValueChangeSetting);
-                [change removeObjectForKey:NSKeyValueChangeIndexesKey];
-            }
+/**
+@Status Interoperable
+*/
+- (void)willChangeValueForKey:(NSString*)key {
+    if ([self observationInfo]) {
+        NSMutableDictionary* change = [NSMutableDictionary dictionaryWithObject:@(NSKeyValueChangeSetting) forKey:NSKeyValueChangeKindKey];
```

Comment:
any reason that this isn't created inside the lambda? (do we want to minimize time inside the lambda or something?)